File: SocketComm/FPipe.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Console;
using System;
using System.Collections;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace SocketComm;

using static Cmd;

public class FPipe : IDisposable, IAsyncDisposable, IEnumerable<Cmd>, IAsyncEnumerable<Cmd>
{
    ILogger<FPipe> Logger { get; }
    private Socket S { get; set; }
    
    private string SFile { get; }

    public event EventHandler Listening;

    public FPipe(string file, ILogger<FPipe>? logger = null)
    {

        this.Logger = logger ?? (LoggerFactory.Create(b => {
            b.AddConsole();
        }).CreateLogger<FPipe>());
        SFile = file;

    }
    public void Connect()
    {
        Logger.LogInformation("Creating socket endpoint");
        UnixDomainSocketEndPoint ep = new UnixDomainSocketEndPoint(SFile);
        Logger.LogInformation("Creating socket");
        if (File.Exists(SFile))
        {
            S = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
            Logger.LogInformation("Connecting...");
            S.Connect(ep);
            ShakeHands(S);
        }
        else
        {
            Socket FileSocket;
            Logger.LogInformation("Serving...");
            FileSocket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
            FileSocket.Bind(ep);
            FileSocket.Listen();
            Logger.LogInformation("Listening");
            Listening?.Invoke(this, EventArgs.Empty);
            S = FileSocket.Accept();
            FileSocket.Disconnect(false);
            FileSocket.Dispose();
            ReceiveHandshake(S);
        }
        Logger.LogInformation("Connected..");
    }
    
    const Int32 MagicHandshakeValue = 25;
    Cmd CmdFromBB(byte[] buffer)
    {
        sbyte result = (sbyte)buffer[0];
        return (Cmd)result;
    }
    byte[] BBFromCmd(Cmd c)
    {
        var b = new byte[] { (byte)c };
        Logger.LogTrace($"{nameof(BBFromCmd)}: {c} ==> {Convert.ToHexString(b)} ({b.Length}b)");
        return b;
    }
    void ValidateHandshake(byte[] data)
    {
        Int32 result = BitConverter.ToInt32(data);
        //L.LogInformation($"[{nameof(ShakeHands)}]: {MagicHandshakeValue}/{result}");
        byte[] rev = new byte[data.Length];
        Array.Copy(data, rev, rev.Length);
        Array.Reverse(rev);
        if (result == MagicHandshakeValue)
        {
            Logger.LogInformation($"Handshake OK {MagicHandshakeValue}/{result}");
            return;
        }

        int revResult = BitConverter.ToInt32(rev);
        if (revResult == MagicHandshakeValue)
        {
            Logger.LogError($"Other side uses {(BitConverter.IsLittleEndian ? "BigEndian" : "LittleEndian")} -- {MagicHandshakeValue}/{result}");
            return;
        }
        throw new InvalidDataException("Bad handshake..." + result);
    }

    public async Task ShakeHandsAsync(Socket socket)
    {
        var data = BitConverter.GetBytes(MagicHandshakeValue);
        await socket.SendAsync(data, SocketFlags.None);
        await socket.ReceiveAsync(data, SocketFlags.None);
        ValidateHandshake(data);
    }

    public void ShakeHands(Socket socket)
    {
        var data = BitConverter.GetBytes(MagicHandshakeValue);
        socket.Send(data, SocketFlags.None);
        socket.Receive(data, SocketFlags.None);
        ValidateHandshake(data);
    }

    public void ReceiveHandshake(Socket socket)
    {
        byte[] data=new byte[4];
        Int32 result;
        Logger.LogInformation("Waiting for handshake");
        socket.Receive(data, SocketFlags.None);
        result = BitConverter.ToInt32(data);
        Logger.LogInformation($"[{nameof(ReceiveHandshake)}] Read: {Convert.ToHexString(data)}/{result}");
        BitConverter.GetBytes(MagicHandshakeValue).CopyTo(data.AsMemory());
        socket.Send(data);
        ValidateHandshake(data);
    }

    public async Task<Cmd> ReadCmdAsync(CancellationToken ct = default)
    {
        try
        {
            var buffer = new byte[1];
            var count = await S.ReceiveAsync(buffer, SocketFlags.None, ct);
            return CmdFromBB(buffer);
        }
        catch(Exception e) {
            Logger.LogError(e.Message);
            return Error; 
        }
    }
    public async Task<bool> WriteCmdAsync(Cmd cmd, CancellationToken cancellationToken = default)
    {
        try
        {
            Logger.LogTrace($"Sending {cmd}");
            var data = BBFromCmd(cmd);
            Logger.LogTrace($"data: {Convert.ToHexString(data)}");
            var sent = await S.SendAsync(data, SocketFlags.None, cancellationToken);
            Logger.LogTrace($"Sent: {sent}");
            return true;
        }
        catch(Exception ex)
        {
            Logger.LogError(ex.Message);
            return false;
        }
    }

    public Cmd ReadCmd()
    {
        try
        {
            byte[] buffer = new byte[1];
            S.Receive(buffer);
            return CmdFromBB(buffer);
        }
        catch (Exception e)
        {
            Logger.LogError(e.Message);
            return Error;
        }
    }

    public Cmd ExpectCmd(Cmd cmd)
    {
        var read = ReadCmd();
        if (cmd == read)
        {
            return read;
        }

        throw new PipeCmdException($"Expected: {cmd} - Received: {read}");
    }

    public async Task<Cmd> ExpectCmdAsync(Cmd cmd, CancellationToken ct = default)
    {
        var read = await ReadCmdAsync(ct);
        if (read == cmd)
        {
            return read;
        }

        throw new PipeCmdException($"Expected: {cmd} - Received: {read}");
    }
    public bool WriteCmd(Cmd cmd)
    {
        try
        {
            S.Send(BBFromCmd(cmd));
            return true;
        }
        catch (Exception e){
            Logger.LogError(e.Message);
            return false;
        }
    }

    public void Dispose()
    {
        Logger.LogInformation("Closing pipe");
        S?.Dispose();
        GC.SuppressFinalize(this);
    }

    public ValueTask DisposeAsync()
    {
        Logger.LogInformation("Closing pipe async");
        S?.Dispose();
        GC.SuppressFinalize(this);
        return ValueTask.CompletedTask;
    }

    public IEnumerator<Cmd> GetEnumerator()
    {
        Cmd r;
        do
        {
            r = ReadCmd();
            yield return r; 
        } while (r > Exit); 
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    public async IAsyncEnumerator<Cmd> GetAsyncEnumerator(CancellationToken cancellationToken = default)
    {
        Cmd r;
        do
        {
            r = await ReadCmdAsync(cancellationToken);
            yield return r;
        } while(r > Exit);
    }

    public void SendValue<T>(T value, Func<T,byte[]> converter)
    {
        WriteCmd(Receive);
        ExpectCmd(Ready);
        var buf = converter(value);
        byte[] length = BitConverter.GetBytes(buf.Length);
        if (BitConverter.IsLittleEndian)
        {
            length = length.Reverse().ToArray();
        }
        S.Send(length);
        S.Send(buf);
        ExpectCmd(Ok);
    }

    public T ReceiveValue<T>(Func<byte[], T> converter)
    {
        WriteCmd(Ready);
        byte[] buffer = new byte[4];
        S.Receive(buffer);
        if (BitConverter.IsLittleEndian)
        {
            buffer = buffer.Reverse().ToArray();
        }
        var length = BitConverter.ToInt32(buffer);
        Console.WriteLine($"Receiving {length} bytes!");
        buffer = new byte[length];
        var rec = S.Receive(buffer);
        if (rec != length)
        {
            //TODO: throw error, maybe try repeat
        }

        WriteCmd(Ok);
        return converter(buffer);
    }
}

File: JSocketComm/src/FPipe.java
import java.io.IOException;
import java.net.StandardProtocolFamily;
import java.net.UnixDomainSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Iterator;
import java.util.concurrent.Callable;
import java.util.function.Function;

public class FPipe implements AutoCloseable, Iterable<Cmd> {
    final int MagicHandshakeValue = 25;
    SocketChannel channel;

    public FPipe(String file) throws IOException {

        UnixDomainSocketAddress address = UnixDomainSocketAddress.of(file);
        if (Files.exists(Path.of(file)))
        {
            channel = SocketChannel.open(StandardProtocolFamily.UNIX);
            channel.connect(address);
            System.out.println("handshake...");
            ShakeHands(channel);
            System.out.println("done");
        } else{
            ServerSocketChannel ss = ServerSocketChannel.open(StandardProtocolFamily.UNIX);
            ss.bind(address);
            channel = ss.accept();
            ss.close();
            System.out.println("handshake...");
            ReceiveHand(channel);
            System.out.println("done");

        }
    }

    public Cmd ReadCmd() throws IOException{
        ByteBuffer bb = ByteBuffer.allocate(1);
        channel.read(bb);
        bb.rewind();
        return Cmd.forValue(bb.get(0));
    }

    public void WriteCmd(Cmd cmd) throws IOException{
        ByteBuffer bb = ByteBuffer.allocate(1);
        channel.write(bb.put(cmd.getValue()).rewind());
    }

    public Cmd ExpectCmd(Cmd cmd) throws PipeCmdException, IOException {
        var read = ReadCmd();
        if(read == cmd){
            return read;
        }
        throw new PipeCmdException(String.format("Expected: %s - Received: %s", cmd, read));
    }

    public <T> void SendValue(T value, Function<T, ByteBuffer> converter) throws IOException, PipeCmdException {
        WriteCmd(Cmd.Receive);
        ExpectCmd(Cmd.Ready);
        var buf = converter.apply(value).rewind();
        ByteBuffer bb = ByteBuffer.allocate(4).putInt(buf.capacity()).rewind();
        channel.write(bb);
        channel.write(buf);
        ExpectCmd(Cmd.Ok);
    }

    public <T> T ReceiveValue(Function<ByteBuffer, T> converter) throws IOException {
        WriteCmd(Cmd.Ready);
        var bb = ByteBuffer.allocate(4);
        channel.read(bb);
        var length = bb.rewind().getInt();
        System.out.println("Receiving " + length + " bytes!");
        bb = ByteBuffer.allocate(length);
        var rec = channel.read(bb);
        if(rec != length){
            // TODO: throw error
        }
        WriteCmd(Cmd.Ok);
        return converter.apply(bb.rewind());
    }


    public void ShakeHands(SocketChannel channel) throws IOException
    {
        ByteBuffer bb = ByteBuffer.allocate(4);
        bb.putInt(MagicHandshakeValue);
        bb.rewind();
        channel.write(bb);
        bb.rewind();
        channel.read(bb);
        bb.rewind();
        int rec = bb.getInt();
        System.out.println("chk: received:"+rec+", expected: "+MagicHandshakeValue);
    }

    public void ReceiveHand(SocketChannel channel) throws IOException{
        ByteBuffer bb = ByteBuffer.allocate(4);
        channel.read(bb);
        int rec = bb.rewind().getInt();
        System.out.println("chk: received:"+rec+", expected: "+MagicHandshakeValue);
        bb.rewind().putInt(MagicHandshakeValue).rewind();
        channel.write(bb);
    }

    @Override
    public void close() throws Exception {
        if(channel!=null)
            channel.close();
    }

    @Override
    public Iterator<Cmd> iterator() {


        return new Iterator<Cmd>() {
            boolean reachedEnd = false;
            Cmd current = Cmd.Ok;
            @Override
            public boolean hasNext(){
                return current.getValue()>Cmd.Exit.getValue();
            }

            @Override
            public Cmd next() {
                try {

                    current = ReadCmd();
                } catch (IOException e){
                    current = Cmd.Error;
                }
                return current;
            }
        };
    }
}